{% from 'macros.html' import danger, info, link %}
{% extends "base.html" %}
{% block more_head %}
	<link rel="stylesheet" href="{{ url_for('static', filename='styles/datatables.min.css') }}">
	<script src="{{ url_for('static', filename='javascript/datatables.min.js') }}" type="text/javascript"></script>	
{% endblock %}
{% block content %}
	<!--button to open left menu-->
	<!--<div class="w3-light-grey" style="position:fixed;width:100%;z-index:2000">-->
	<!--	<span class="w3-button w3-blue w3-xlarge" id="openNav" onclick="w3_open()" style="visibility:visible" title="Click here to open the menu of useful external links">-->
	<!--		<i class="fa fa-bars w3-xlarge"></i>-->
	<!--	</span>-->
	<!--	<span class="w3-button w3-blue w3-xlarge" id="ExportPdf" style="visibility:visible" title="Click here to export the tables as pdf file">-->
	<!--		<i class="fa fa-file-pdf-o w3-xlarge"></i>-->
	<!--	</span>-->
	<!--	<span class="w3-xlarge w3-text-blue w3-button" title="click to get somewhere" onclick="document.getElementById('gene_modal').style.display='block';"><em>{{ variant_features.gene_name[0] }}:{{ variant_features.gene_name[1] }}.{{ variant_features.nm_version }}:c.{{ variant_features.c_name }}</em></span>-->
	<!--</div>-->
	<!--left menu-->
	<div class="w3-sidebar w3-bar-block w3-card w3-animate-left w3-light-grey" id="smart_menu" style="display:none;z-index:2001;width:15%;">
		<span class="w3-bar-item w3-button w3-large w3-border-bottom" onclick="w3_close()">Close Ã—</span>
		{% set hg19_var = namespace(pos='', ref='', alt='', g_name='') %}
		{% set hg38_var = namespace(pos='', ref='', alt='', g_name='') %}
		{% for vari in variant %}			
			{% if vari.genome_version == 'hg19' %}
				{# set variables #}
				{% set hg19_var.pos = vari.pos %}
				{% set hg19_var.ref = vari.pos_ref %}
				{% set hg19_var.alt = vari.pos_alt %}
				{% set hg19_var.g_name = vari.g_name %}
				{# deal with position for genome browsers #}
				{% set positions = namespace(min=vari.pos-25, max=vari.pos+25, end=vari.pos) %}
				{% if variant_features.variant_size > 1 %}
					{% set positions.end = vari.pos + variant_features.variant_size %}
					{% set positions.max = vari.pos + variant_features.variant_size +25 %}
				{% endif %}
				{{ link("http://evs.gs.washington.edu/EVS/PopStatsServlet?searchBy=chromosome&chromosome=%s&chromoStart=%s&chromoEnd=%s&x=0&y=0"|format(vari.chr, vari.pos, positions.end), "ESP6500") }}
				{{ link("http://gnomad.broadinstitute.org/variant/%s-%s-%s-%s"|format(vari.chr, vari.pos, vari.pos_ref, vari.pos_alt), "gnomAD")}}
				{{ link("http://wintervar.wglab.org/results.pos.php?queryType=position&build=hg19_updated.v.201904&chr=%s&pos=%s&ref=%s&alt=%s"|format(vari.chr, vari.pos, vari.pos_ref, vari.pos_alt), "InterVar")}}
				{{ link("https://www.ncbi.nlm.nih.gov/variation/tools/1000genomes/?chr=%s&from=%s&to=%s&gts=rs%s&mk=%s:%s|rs%s"|format(vari.chr, positions.min, positions.max, variant_features.dbsnp_id, vari.pos, vari.pos, variant_features.dbsnp_id), "1000 genomes")}}
				{{ link("http://www.ncbi.nlm.nih.gov/clinvar?term='%s.%s:c.%s' [Variant name]"|format(variant_features.gene_name[1], variant_features.nm_version, variant_features.c_name), "Clinvar")}}
				{{ link("http://www.ncbi.nlm.nih.gov/snp/rs%s"|format(variant_features.dbsnp_id), "dbSNP")}}
				{{ link("http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr%s:%s-%s&highlight=hg19.chr%s:%s-%s&hgsid=757149165_H4Gy8jWA4BwCuA6WW1M8UxC37MFn"|format(vari.chr, positions.min, positions.max, vari.chr, vari.pos, positions.end), "hg19 UCSC")}}
				{{ link("http://www.rcsb.org/pdb/chromosome.do?v=hg37&chromosome=%s&pos=%s"|format(vari.chr, vari.pos), "hg19 Map2PDB") }}
				{{ link("https://varsome.com/variant/hg19/%s.%s:c.%s"|format(variant_features.gene_name[1], variant_features.nm_version, variant_features.c_name), "VarSome") }}
			{% elif vari.genome_version == 'hg38' %}
				{# set variables #}
				{% set hg38_var.pos = vari.pos %}
				{% set hg38_var.ref = vari.pos_ref %}
				{% set hg38_var.alt = vari.pos_alt %}
				{% set hg38_var.g_name = vari.g_name %}
				{# deal with position for genome browsers #}
				{% set positions = namespace(min=vari.pos-25, max=vari.pos+25, end=vari.pos) %}
				{% if variant_features.variant_size > 1 %}
					{% set positions.end = vari.pos + variant_features.variant_size %}
					{% set positions.max = vari.pos + variant_features.variant_size +25 %}
				{% endif %}
				{{ link("http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr%s:%s-%s&highlight=hg38.chr%s:%s-%s&hgsid=757151741_tUQRsEoYvqXsPekoiKpwY6Ork8eF"|format(vari.chr, positions.min, positions.max, vari.chr, vari.pos, positions.end), "hg38 UCSC") }}
				{{ link("http://www.rcsb.org/pdb/chromosome.do?v=hg38&chromosome=%s&pos=%s"|format(vari.chr, vari.pos), "hg38 Map2PDB") }}
			{% endif %}		
		{% endfor %}
	</div><br />
	<!--title-->
	<!--<div class="w3-center">-->
	<!--	<h1 class="w3-text-blue w3-button w3-hover-sand" title="click to get somewhere" onclick="document.getElementById('gene_modal').style.display='block';"><em>{% block title %}{{ variant_features.gene_name[0] }}:{{ variant_features.gene_name[1] }}.{{ variant_features.nm_version }}:c.{{ variant_features.c_name }}{% endblock %}</em></h1>		-->
	<!--</div>-->
	<br />
	<!--first section-->
	<hr />
	<div class="w3-text-red w3-container">
		<h2 id="nom">Nomenclatures</h2>
	</div>
	<div class="w3-responsive w3-container">
		<table class="w3-table w3-centered display compact" style="width:90%" id="nomenclature_table">
			<caption>Variant details:</caption>
			<thead>
				<tr>
					<th class="w3-left-align">Features</th>
					<th class="w3-left-align">Values</th>
					<th class="w3-left-align">Description</th>
				</tr>
			</thead>
			<tbody>
				<!--HGVS DNA on transcript-->
				<tr>
					<td class="w3-left-align">HGVS DNA on transcript:</td>
					<td class="w3-left-align"><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ variant_features.gene_name[1] }}.{{ variant_features.nm_version }}" target="_blank" title="open NCBI page for {{ variant_features.gene_name[1] }}.{{ variant_features.nm_version }}">{{ variant_features.gene_name[1] }}.{{ variant_features.nm_version }}</a>:<a href="https://mutalyzer.nl/name-checker?description={{ variant_features.gene_name[1] }}.{{ variant_features.nm_version }}:c.{{ variant_features.c_name }}"  target="_blank" title="Check Mutalyzer for the variant">c.{{ variant_features.c_name }}</a></td>
					<td class="w3-left-align">HGVS full nomenclature at DNA level on transcript</td>
				</tr>
				<!--IVS name-->
				{% if variant_features.start_segment_type == 'intron' %}
					<tr>
						<td class="w3-left-align">IVS name:</td>
						<td class="w3-left-align">{{ variant_features.ivs_name }}</td>
						<td class="w3-left-align">IVS nomenclature</td>
					</tr>
				{% endif %}				
				<!--HGVS Protein-->
				<tr>
					<td class="w3-left-align">HGVS Protein:</td>
					<td class="w3-left-align"><a href="https://www.ncbi.nlm.nih.gov/protein/{{ variant_features.np }}" target="_blank" title="open NCBI page for {{ variant_features.np }}">{{ variant_features.np }}</a>:p.({{ variant_features.p_name }})</td>
					<td class="w3-left-align">HGVS full nomenclature at protein level</td>
				</tr>
				<!--HGVS genomic (gene)-->
				<tr>
					<td class="w3-left-align">HGVS genomic (gene):</td>
					<td class="w3-left-align"><a href="https://www.ncbi.nlm.nih.gov/nuccore/{{ variant_features.ng }}" target="_blank" title="open NCBI page for {{ variant_features.ng }}">{{ variant_features.ng }}</a>:g.{{ variant_features.ng_name }}</td>
					<td class="w3-left-align">HGVS full nomenclature at genomic level (gene centered)</td>
				</tr>
				<!--HGVS genomic (hg19)-->
				<tr>
					<td class="w3-left-align">HGVS genomic (hg19):</td>
					<td class="w3-left-align">hg19:{{ hg19_var.g_name }}</td>
					<td class="w3-left-align">HGVS full nomenclature at genomic level (hg19)</td>
				</tr>
				<!--pseudo VCF (hg19)-->
				<tr>
					<td class="w3-left-align">pseudo VCF (hg19):</td>
					<td class="w3-left-align"><a href="https://variantvalidator.org/variantvalidation/?variant={{ variant_features.chr }}-{{ hg19_var.pos }}-{{ hg19_var.ref }}-{{ hg19_var.alt }}&primary_assembly=GRCh37&alignment=splign" target="_blank" title="open VariantValidator">hg19:{{ variant_features.chr }}-{{ hg19_var.pos }}-{{ hg19_var.ref }}-{{ hg19_var.alt }}</a></td>
					<td class="w3-left-align">chr-pos-ref-alt (hg19)</td>
				</tr>
				<!--HGVS genomic (hg38)-->
				<tr>
					<td class="w3-left-align">HGVS genomic (hg38):</td>
					<td class="w3-left-align">hg38:{{ hg38_var.g_name }}</td>
					<td class="w3-left-align">HGVS full nomenclature at genomic level (hg38)</td>
				</tr>
				<!--pseudo VCF (hg38)-->
				<tr>
					<td class="w3-left-align">pseudo VCF (hg38):</td>
					<td class="w3-left-align"><a href="https://variantvalidator.org/variantvalidation/?variant={{ variant_features.chr }}-{{ hg38_var.pos }}-{{ hg38_var.ref }}-{{ hg38_var.alt }}&primary_assembly=GRCh38&alignment=splign" target="_blank" title="open VariantValidator">hg38:{{ variant_features.chr }}-{{ hg38_var.pos }}-{{ hg38_var.ref }}-{{ hg38_var.alt }}</a></td>
					<td class="w3-left-align">chr-pos-ref-alt (hg38)</td>
				</tr>
			</tbody>
		</table><br/><br/>
		<div class = "w3-row">
			<div class = "w3-col m4">
				<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="" style="width:100%">Mutalyzer</span>
			</div>
			<div class = "w3-col m4">
				<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="" style="width:100%">VariantValidator</span>
			</div>
			<div class = "w3-col m4">
				<span class="w3-button w3-blue w3-ripple w3-hover-light-blue w3-padding-16 w3-large" onclick="" style="width:100%">HGVS website</span>
			</div>
		</div>
	</div>
	<hr />
	<div class="w3-text-red w3-container">
		<h2 id="nom">Positions</h2>
	</div>
	<div class="w3-responsive w3-container">
		<!--Position in transcript-->
		<table class="w3-table w3-centered display compact" style="width:90%" id="position_table">
			<caption>Variant details:</caption>
			<thead>
				<tr>
					<th class="w3-left-align">Features</th>
					<th class="w3-left-align">Values</th>
					<th class="w3-left-align">Description</th>
				</tr>
			</thead>
			<tbody>
				<tr>
					<td class="w3-left-align">Position in transcript:</td>					
					{% if variant_features.start_segment_type == variant_features.end_segment_type and variant_features.start_segment_number == variant_features.end_segment_number %}
						<td class="w3-left-align">{{ variant_features.start_segment_type|capitalize }} {{ variant_features.start_segment_number }}</td>
					{% else %}
						<td class="w3-left-align">{{ variant_features.start_segment_type|capitalize }} {{ variant_features.start_segment_number }} - {{ variant_features.end_segment_type|capitalize }} {{ variant_features.end_segment_number }}</td>
					{% endif %}
					<td class="w3-left-align">Exon/intron position in {{ variant_features.gene_name[1] }}.{{ variant_features.nm_version }}</td>
				</tr>
			</tbody>
		</table><br/><br/>
	</div>
	
	<br/><br/>
	<script>
		$(document).ready(function() {
			$('.w3-table').DataTable({
				responsive: true,
				dom: 'ft',
				"order": [],
				//scrollY: 600,
				buttons: [
					'copy', 'excel', 'pdf'
				]
			});
			//adapted from https://sharepoint.stackexchange.com/questions/234464/datatables-plugin-print-multiple-tables-on-one-page
			// export multiple tables in one single pdf
			$('#ExportPdf').click(function(){

				var config =    {
					className:"buttons-pdf buttons-html5",
					customize:null,
					download:"download",
					exportOptions:{},
					extension:".pdf",
					filename:"*",
					footer:false,
					header:true,
					messageBottom:"*",
					messagetop:"*",
					namespace:".dt-button-2",
					orientation:"portrait",
					pageSize:"A4",
					title:"*"
				};
	
				var tables = ["nomenclature_table","position_table"];
				var tablesConverted = {};
				for(var k=0;k<tables.length;k++)
				{
					var dt = $('#'+tables[k]).DataTable();
					var data = dt.buttons.exportData( config.exportOptions );
					var info = dt.buttons.exportInfo( config );
	
					var rows = [];
	
						if ( config.header ) {
							rows.push( $.map( data.header, function ( d ) {
								return {
									text: typeof d === 'string' ? d : d+'',
									style: 'tableHeader'
								};
							} ) );
						}
	
						for ( var i=0, ien=data.body.length ; i<ien ; i++ ) {
							rows.push( $.map( data.body[i], function ( d ) {
								return {
									text: typeof d === 'string' ? d : d+'',
									style: i % 2 ? 'tableBodyEven' : 'tableBodyOdd'
								};
							} ) );
						}
	
						if ( config.footer && data.footer) {
							rows.push( $.map( data.footer, function ( d ) {
								return {
									text: typeof d === 'string' ? d : d+'',
									style: 'tableFooter'
								};
							} ) );
						}
	
						tablesConverted[tables[k]]=rows;
				}
	
	
				var doc = {
					pageSize: config.pageSize,
					pageOrientation: config.orientation,
					content: [
						"Data for "+tables[0],
						" ",
						{
							table: {
								headerRows: 1,
								body: tablesConverted[tables[0]]
							},
							layout: 'noBorders'
						},
						" ",
						"Data for "+tables[1],
						" ",
						{
							table: {
								headerRows: 1,
								body: tablesConverted[tables[1]]
							},
							layout: 'noBorders'
						},
						" ",
						//"Data for "+tables[2],
						//" ",
						//{
						//	table: {
						//		headerRows: 1,
						//		body: tablesConverted[tables[2]]
						//	},
						//	layout: 'noBorders'
						//},
					],
					styles: {
						tableHeader: {
							bold: true,
							fontSize: 11,
							color: 'white',
							fillColor: '#2d4154',
							alignment: 'center'
						},
						tableBodyEven: {},
						tableBodyOdd: {
							fillColor: '#f3f3f3'
						},
						tableFooter: {
							bold: true,
							fontSize: 11,
							color: 'white',
							fillColor: '#2d4154'
						},
						title: {
							alignment: 'center',
							fontSize: 15
						},
						message: {}
					},
					defaultStyle: {
						fontSize: 10
					}
				};

				if ( info.messageTop ) {
					doc.content.unshift( {
						text: info.messageTop,
						style: 'message',
						margin: [ 0, 0, 0, 12 ]
					} );
				}

				if ( info.messageBottom ) {
					doc.content.push( {
						text: info.messageBottom,
						style: 'message',
						margin: [ 0, 0, 0, 12 ]
					} );
				}

				if ( info.title ) {
					doc.content.unshift( {
						text: info.title,
						style: 'title',
						margin: [ 0, 0, 0, 12 ]
					} );
				}

				if ( config.customize ) {
					config.customize( doc, config );
				}	
				{% autoescape false %}
				pdfMake.createPdf(doc).download('{{ variant_features.gene_name[1] }}.{{ variant_features.nm_version }}_{{ variant_features.c_name }}.pdf');
				{% endautoescape %}
			});
		} );
	</script>

{% endblock %}
